components/config/prefs.rs

```diff
@@ -7,251 +7,398 @@ use crate::opts;
 use embedder_traits::resources::{self, Resource};
 use serde_json::{self, Value};
 use std::borrow::ToOwned;
-use std::cmp::max;
 use std::collections::HashMap;
 use std::fs::File;
 use std::io::{stderr, Read, Write};
 use std::path::PathBuf;
-use std::sync::{Arc, RwLock};
+
+use crate::pref_util::Preferences;
+pub use crate::pref_util::{PrefError, PrefValue};
+use gen::Prefs;
 
 lazy_static! {
-    pub static ref PREFS: Preferences = {
-        let defaults = default_prefs();
-        if let Ok(prefs) = read_prefs(&resources::read_string(Resource::Preferences)) {
-            defaults.extend(prefs);
-        }
-        defaults
+    static ref PREFS: Preferences<'static, Prefs> = {
+        let def_prefs: Prefs = serde_json::from_str(&resources::read_string(Resource::Preferences))
+            .expect("Failed to initialize config preferences.");
+        Preferences::new(def_prefs, &gen::PREF_ACCESSORS)
     };
 }
 
-#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
-pub enum PrefValue {
-    Boolean(bool),
-    String(String),
-    Number(f64),
-    Missing,
-}
-
-impl PrefValue {
-    pub fn from_json(data: Value) -> Result<PrefValue, ()> {
-        let value = match data {
-            Value::Bool(x) => PrefValue::Boolean(x),
-            Value::String(x) => PrefValue::String(x),
-            Value::Number(x) => {
-                if let Some(v) = x.as_f64() {
-                    PrefValue::Number(v)
-                } else {
-                    return Err(());
-                }
-            },
-            _ => return Err(()),
-        };
-        Ok(value)
-    }
-
-    pub fn as_boolean(&self) -> Option<bool> {
-        match *self {
-            PrefValue::Boolean(value) => Some(value),
-            _ => None,
-        }
-    }
-
-    pub fn as_string(&self) -> Option<&str> {
-        match *self {
-            PrefValue::String(ref value) => Some(&value),
-            _ => None,
-        }
-    }
-
-    pub fn as_i64(&self) -> Option<i64> {
-        match *self {
-            PrefValue::Number(x) => Some(x as i64),
-            _ => None,
-        }
-    }
-
-    pub fn as_u64(&self) -> Option<u64> {
-        match *self {
-            PrefValue::Number(x) => Some(x as u64),
-            _ => None,
-        }
-    }
-}
-
-#[derive(Clone, Debug, Deserialize, Serialize)]
-pub enum Pref {
-    NoDefault(Arc<PrefValue>),
-    WithDefault(Arc<PrefValue>, Option<Arc<PrefValue>>),
+/// A convenience macro for accessing a preference value using its static path.
+/// Passing an invalid path is a compile-time error.
+#[macro_export]
+macro_rules! get_pref {
+    ($($segment: ident).+) => {{
+        let values = $crate::prefs::pref_map().values();
+        let lock = values.read()
+            .map(|prefs| prefs $(.$segment)+.clone());
+        lock.unwrap()
```

Comment:
Nit: do we need to `clone` here?

Reply:
There might be a better design. It's needed because `String` is one of the supported preference types, not `&str`.

Reply:
But actually, it's not so simple. To keep the macro self-contained, the lock has to be dropped inside the macro expansion. It's not possible to return a reference from here. If it is important to not clone then users of a string pref have to own the lock themselves and not use this macro.

Reply:
Something like this also won't work:
```
macro_rules! use_pref {
    ($($segment: ident).+, $closure: expr) => {{
        let values = $crate::prefs::pref_map().values();
        let lock = values.read();
        lock.as_ref()
            .map(|prefs| $closure(&prefs $(.$segment)+))
            .unwrap()
    }};
}
```
Because you _could_ call `pref!` or `set_pref!` inside the closure, which would cause deadlock.
```
let home = "https://servo.org";
// This is OK
assert_eq!(true, use_pref!(shell.homepage, |url| url == home));

use_pref!(shell.homepage, |url| if url != home {
    // DEADLOCK
    set_pref!(shell.homepage, home.to_string());
});
```

Reply:
Oh good point. That's not so much of an issue if there aren't any uses of string prefs inside a hot loop. The alternative would be to make the API something like `prefs::lookup_with(key, callback)`, which isn't as nice but avoids cloning strings.